kube-controllers/pkg/controllers/node/hostendpoints.go

```diff
@@ -17,300 +17,425 @@ package node
-				if c.config.AutoHostEndpoints && c.syncStatus == bapi.InSync {
-					hepName := c.generateAutoHostendpointName(nodeName)
-					err := c.deleteHostendpointWithRetries(context.Background(), hepName)
-					if err != nil {
-						logrus.WithError(err).Fatal()
-					}
-				}
-			}
+	}
+
+}
+
+func (c *autoHostEndpointController) handleUpdate(update interface{}) {
+	switch update := update.(type) {
+	case bapi.SyncStatus:
+		c.syncStatus = update
+		switch update {
+		case bapi.InSync:
+			logrus.WithField("status", update).Info("Syncer is InSync, kicking sync channel")
+			kick(c.syncChan)
+		}
+		return
+	case model.KVPair:
+		switch update.Key.(model.ResourceKey).Kind {
+		case libapi.KindNode:
+			c.handleNodeUpdate(update)
+			return
+		case api.KindHostEndpoint:
+			c.handleHostEndpointUpdate(update)
+			return
 		}
 	}
 }
 
-// deleteAutoHostendpointsWithoutNodes deletes auto hostendpoints that either
-// reference a Calico node that doesn't exist, or, that remain after
-// autoHostEndpoints has been disabled.
-func (c *autoHostEndpointController) deleteAutoHostendpointsWithoutNodes(ctx context.Context, heps map[string]api.HostEndpoint) error {
-	for _, hep := range heps {
-		_, hepNodeExists := c.nodeCache[hep.Spec.Node]
+// handleNodeUpdate handles node updates received via the syncer.
+// On delete and update we want to trigger hostEndpointSync for this node
+func (c *autoHostEndpointController) handleNodeUpdate(kvp model.KVPair) {
+	nodeName := kvp.Key.(model.ResourceKey).Name
+	if kvp.Value == nil {
+		// Node deleted, we want to remove all auto HostEndpoints associated with this node
+		delete(c.nodeCache, nodeName)
+	} else {
+		node := kvp.Value.(*libapi.Node)
+		c.nodeCache[node.Name] = node
+	}
+
+	if c.syncStatus == bapi.InSync {
+		c.nodeUpdates <- nodeName
+	}
+}
 
-		if !hepNodeExists || !c.config.AutoHostEndpoints {
-			err := c.deleteHostendpoint(ctx, hep.Name)
+// handleHostEndpointUpdate handles HostEndpoint updates recieved via syncer.
+// We want to delete HostEndpoints that no longer exists, and add/update Auto-HostEndpoints to our local cache
+func (c *autoHostEndpointController) handleHostEndpointUpdate(kvp model.KVPair) {
+	if kvp.Value == nil {
+		c.hostEndpointTracker.deleteHostEndpointForNode(kvp.Key.(model.ResourceKey).Name)
+		return
+	}
+
+	hostEndpoint := kvp.Value.(*api.HostEndpoint)
+
+	if isAutoHostEndpoint(hostEndpoint) {
+		c.hostEndpointTracker.addHostEndpointForNode(hostEndpoint)
+	}
+}
+
+// syncHostEndpoints() is the main function that is responsible for keeping HostEndpoints in sync for nodes. It does the following:
+// 1. Checks if the hostEndpoint controller is disabled, in that case we should delete all previously created HostEndpoints
+// 2. Clean up any HostEndpoints that are lingering after a node was deleted
+// 3. Go through each node and sync HostEndpoints for that node
+func (c *autoHostEndpointController) syncHostEndpoints() {
+	logrus.Infof("Syncing all HostEndpoints")
```

Comment:
can move this below the InSync check below, otherwise the log output will be misleading.